compiler/rustc_ast/src/ast.rs

```diff
@@ -3382,25 +3383,38 @@ pub struct PolyTraitRef {
     /// The `Foo<&'a T>` in `<'a> Foo<&'a T>`.
     pub trait_ref: TraitRef,
 
+    /// The source of this trait ref
+    /// and whether this trait ref should be regarded as
+    /// a nominal subtrait-supertrait relation.
+    pub source: TraitRefSource,
```

Comment:
We probably don't need this in AST.
This flag is needed after HIR->Ty lowering, I guess, and instead of parser it's equally simple to calculate it at any stage before that (e.g. in AST->HIR lowering, or HIR->Ty lowering).

Reply:
How about switching out `parse_generic_bounds` parser in `fn parse_item_trait` and we record the supertrait modifier on the trait item level?

Reply:
The `auto impl` modifier is in AST, yes.
Parser generally shouldn't be too context-sensitive and use things like `in_supertrait_context`, so it would typically parse `auto impl` everywhere and then later passes would reject it semantically in inappropriate positions.
Otherwise we have two separate but similar grammars for bound lists.